src/client/renderingengine.cpp

```diff
@@ -65,6 +66,45 @@ static gui::GUISkin *createSkin(gui::IGUIEnvironment *environment,
 }
 
 
+static std::optional<video::E_DRIVER_TYPE> chooseVideoDriver()
+{
+	auto &&configured_name = g_settings->get("video_driver");
+	if (configured_name.empty())
+		return std::nullopt;
+
+	auto &&drivers = RenderingEngine::getSupportedVideoDrivers();
+	for (auto driver: drivers) {
+		auto &&info = RenderingEngine::getVideoDriverInfo(driver);
+		if (!strcasecmp(configured_name.c_str(), info.name.c_str()))
+			return driver;
+	}
+
+	errorstream << "Invalid video_driver specified: " << configured_name << std::endl;
+	return std::nullopt;
+}
+
+static irr::IrrlichtDevice *createDevice(SIrrlichtCreationParameters params, std::optional<video::E_DRIVER_TYPE> requested_driver)
+{
+	if (requested_driver) {
+		params.DriverType = *requested_driver;
+		if (auto *device = createDeviceEx(params))
+			return device;
+		errorstream << "Failed to initialize the " << RenderingEngine::getVideoDriverInfo(*requested_driver).friendly_name << " video driver" << std::endl;
+	}
+	sanity_check(requested_driver != video::EDT_NULL);
```

Comment:
Just a quick question: This check is because the NULL driver may never fail, right?

Reply:
Yes, and also because if it does (e.g. compiled out), falling back to anything else doesn’t make much sense.